src/transformers/models/beit3/modeling_beit3.py

```diff
@@ -0,0 +1,1204 @@
+# coding=utf-8
+# Copyright 2023 Microsoft Research and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" PyTorch BEiT3 model."""
+
+
+import copy
+import math
+from dataclasses import dataclass
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+from torch import nn
+from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, LayerNorm, MSELoss
+
+from transformers import PreTrainedModel, add_start_docstrings
+from transformers.activations import get_activation
+from transformers.modeling_outputs import (
+    CausalLMOutputWithPast,
+    ImageClassifierOutputWithNoAttention,
+    SequenceClassifierOutput,
+)
+from transformers.models.beit3.configuration_beit3 import Beit3Config
+from transformers.utils import ModelOutput, add_start_docstrings_to_model_forward, logging
+
+
+logger = logging.get_logger(__name__)
+
+
+BEIT3_START_DOCSTRING = r"""
+    This model is a PyTorch [torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) subclass. Use it
+    as a regular PyTorch Module and refer to the PyTorch documentation for all matter related to general usage and
+    behavior.
+
+    Parameters:
+        config ([`Beit3Config`]): Model configuration class with all the parameters of the model.
+            Initializing with a config file does not load the weights associated with the model, only the
+            configuration. Check out the [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+
+BEIT3_FOR_VISUALREASONING_INPUTS_DOCSTRING = r"""
+    Args:
+        input_ids (`torch.LongTensor` of shape `({0})`):
+            Indices of input sequence tokens in the vocabulary.
+
+            Indices can be obtained using [`AutoTokenizer`]. See [`PreTrainedTokenizer.encode`] and
+            [`PreTrainedTokenizer.__call__`] for details.
+
+            [What are input IDs?](../glossary#input-ids)
+        image1_values (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Pixel values. Pixel values can be obtained using [`AutoImageProcessor`]. See
+            [`BeitImageProcessor.__call__`] for details.
+        image2_values (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Pixel values. Pixel values can be obtained using [`AutoImageProcessor`]. See
+            [`BeitImageProcessor.__call__`] for details.
```

Comment:
FYI we already have [ViLT](https://github.com/huggingface/transformers/blob/03e309d58ec1d5dcb3083bf8e78ace1b389af421/src/transformers/models/vilt/modeling_vilt.py#L1281) which also does visual reasoning on NLVR, which just takes `pixel_values` as input, being a tensor of shape (batch_size, 2, num_channels, height, width) - since there are 2 images for every training example. Maybe we could do the same here?

Reply:
Oh, thats a great idea, I will follow the same approach here too .